File: app/src/main/java/com/example/place/MainActivity.java
package com.example.place;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import android.content.DialogInterface;
import android.content.Intent;
import android.hardware.Sensor;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.Gravity;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import com.example.settlingmeasurement.Sensing;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

public class MainActivity extends AppCompatActivity {

    protected static final String TAG = MainActivity.class.getSimpleName();
    private Quiz quiz;  //英単語問題データクラス
    private  int numOfCorrect;  //正解数
    AlertDialog alertDialog;
    MetaData metaData = MetaData.getInstance();


    @RequiresApi(api = Build.VERSION_CODES.O)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar myToolbar = (Toolbar) findViewById(R.id.toolBar);
        setSupportActionBar(myToolbar);


        Button unknownBtn = findViewById(R.id.unknown_btn);
        Button rememberBtn = findViewById(R.id.remember_btn);
        Button startBtn = findViewById(R.id.start_btn);
        Button vocabBtn = findViewById(R.id.vocabulary_btn);
        Button metaBtn = findViewById(R.id.meta_btn);


        numOfCorrect = 0;
        quiz = new Quiz();
        this.UpdateMemory();//まだよくわからない


        //記憶数によって表示を変化させる
        int unknownWords = quiz.getQuizData().length - numOfCorrect;
        rememberBtn.setText("Remember\n"+numOfCorrect);
        unknownBtn.setText("Unknown\n"+unknownWords);



        startBtn.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view){
                //問題画面へ遷移
                //メタデータを選択しないと遷移しないようにする
                if(metaData.getQuizPattern() != null){
                    Intent qIntent = new Intent(getApplicationContext(), QuestionActivity.class);
                    startActivity(qIntent);
                }else{
                    Toast toast = Toast.makeText(getApplicationContext(),"メタデータを入力して下さい",Toast.LENGTH_LONG);
                    // 位置調整
                    toast.setGravity(Gravity.CENTER, 0, -200);
                    toast.show();
                }

                Log.d(TAG, "onClick:start_btn");
            }
        });

        vocabBtn.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view){
                //MyWords画面へ遷移
                Intent vocabIntent = new Intent(getApplicationContext(), VocabActivity.class);
                startActivity(vocabIntent);
                Log.d(TAG, "onClick:vocab_btn");
            }
        });


        metaBtn.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view){
                //メタデータ入力画面へ遷移
                Intent inputMetaIntent = new Intent(getApplicationContext(), InputNoteActivity.class);
                startActivity(inputMetaIntent);
                Log.d(TAG, "onClick:meta_btn");
            }
        });


    }
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.toolbar, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.config:
                //設定画面へ遷移
                Intent configIntent = new Intent(getApplicationContext(), ConfigActivity.class);
                startActivity(configIntent);
                Log.d(TAG, "onClick:config_btn");
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    //戻るでアプリ終了
    @Override
    public void onBackPressed() {
        alertDialog = new AlertDialog.Builder(MainActivity.this)
                .setCancelable(false)
                .setTitle("確認")
                .setMessage("終了してよろしいですか")
                .setPositiveButton("Yes", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        finish();
                    }
                })
                .setNegativeButton("No", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        alertDialog.dismiss();
                    }
                }).show();
    }




    public String readFile(String file){
        String text = null;

        try (FileInputStream fileInputStream = openFileInput(file);
             BufferedReader reader= new BufferedReader(
                     new InputStreamReader(fileInputStream, StandardCharsets.UTF_8))) {

            String lineBuffer;
            while( (lineBuffer = reader.readLine()) != null ) {
                text = lineBuffer ;
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
        return text;
    }


    //内部ストレージに入っている記憶データを用いてフラグを更新
    public void UpdateMemory() {
        //内部ストレージに存在しているかで分岐
        String text = readFile("MyMemory");
        if (text == null) {
        } else {
            String[] temp = text.split(",", quiz.getQuizData().length);
            for (int i = 0; i < quiz.getQuizData().length; i++) {
                if (temp[i].equals("1")) {
                    quiz.setMemory(i, true);
                    numOfCorrect++;
                }
            }
        }
    }

}

File: app/src/main/java/com/example/place/activity/MainActivity.kt
package com.example.place.activity

import android.annotation.SuppressLint
import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.*
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.AlarmManagerCompat
import androidx.preference.PreferenceManager
import com.example.place.MetaData.Companion.getInstance
import com.example.place.R
import com.example.place.databinding.ActivityMainBinding
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*

class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding
    private var alertDialog: AlertDialog? = null
    private var metaData = getInstance()
    private var measurementTime:Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val myToolbar = binding.toolBar
        setSupportActionBar(myToolbar)
        val mAuth = FirebaseAuth.getInstance()

        val currentUser = mAuth.currentUser
        val currentUserTxt = binding.currentUserTxt
        currentUserTxt.text = String.format("%s\nでログインしています．",currentUser!!.email)

        val startBtn = binding.startBtn
        val metaBtn = binding.metaBtn

        measurementTime = PreferenceManager.getDefaultSharedPreferences(this).getInt("measurementTime",10)


        startBtn.setOnClickListener { //問題画面へ遷移
            //メタデータを選択しないと遷移しないようにする
            if (metaData.quizPattern != null || metaData.labelData != null) {
                startMeasurementAlarm()

                val intent = if(PreferenceManager.getDefaultSharedPreferences(this).getBoolean("flashCardMode",false)){
                    Intent(applicationContext, FlashCardActivity::class.java)
                }else{
                    Intent(applicationContext, QuestionActivity::class.java)
                }
                startActivity(intent)
            } else {
                Snackbar.make(binding.root,"メタデータ入力してください", Snackbar.LENGTH_SHORT)
                    .setAction("メタデータ入力") {
                        val inputMetaIntent =
                            Intent(applicationContext, InputNoteActivity::class.java)
                        startActivity(inputMetaIntent)
                    }
                    .show()
            }
            Log.d(TAG, "onClick:start_btn")
        }

        metaBtn.setOnClickListener { //メタデータ入力画面へ遷移
            val inputMetaIntent = Intent(applicationContext, InputNoteActivity::class.java)
            startActivity(inputMetaIntent)
            Log.d(TAG, "onClick:meta_btn")
        }
    }

    //10分計測のためのアラームをセットする
    @SuppressLint("UnspecifiedImmutableFlag")
    private fun startMeasurementAlarm() {

        // 時間をセットする
        val calendar = Calendar.getInstance()
        // Calendarを使って現在の時間をミリ秒で取得
        calendar.timeInMillis = System.currentTimeMillis()
        // metadataに保存されている分数後に設定
        calendar.add(Calendar.SECOND, measurementTime * 60)

        //時間精度デバック用
        val df: DateFormat = SimpleDateFormat("yyyy/MM/dd HH:mm:ss.SSS", Locale.JAPAN)
        val date = Date(System.currentTimeMillis())
        Log.d("alarmCheck_start", df.format(date))

        //暗黙的なBroadCast
        val intent = Intent("STOP")
        val pending = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            PendingIntent.getBroadcast(
                applicationContext, 0, intent, PendingIntent.FLAG_IMMUTABLE
            )}else{
                PendingIntent.getBroadcast(
                    applicationContext, 0, intent, PendingIntent.FLAG_CANCEL_CURRENT)
            }

        // アラームをセットする
        val am = getSystemService(ALARM_SERVICE) as AlarmManager
        //                        am.setExact(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis(), pending);
        AlarmManagerCompat.setExact(am, AlarmManager.RTC_WAKEUP, calendar.timeInMillis, pending)
        Toast.makeText(
            applicationContext,
            "${measurementTime}分の計測を始めます.", Toast.LENGTH_SHORT
        ).show()
    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.toolbar, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.config) { //設定画面へ遷移
            val configIntent = Intent(applicationContext, ConfigActivity::class.java)
            startActivity(configIntent)
            Log.d(TAG, "onClick:config_btn")
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    //戻るでアプリ終了
    override fun onBackPressed() {
        AlertDialog.Builder(this@MainActivity)
            .setCancelable(false)
            .setTitle("確認")
            .setMessage("終了してよろしいですか")
            .setPositiveButton("Yes") { _, _ -> finish() }
            .setNegativeButton("No") { _, _ ->  }.show()
    }

    companion object{
        private const val TAG = "MainActivity"
    }
}